marklogic-data-hub/src/main/java/com/marklogic/hub/step/impl/QueryStepRunner.java

```diff
@@ -524,4 +513,23 @@ private RunStepResponse runHarmonizer(RunStepResponse runStepResponse, Collectio
         runningThread.start();
         return runStepResponse;
     }
+
+    private String determineStepStatus(StepMetrics stepMetrics) {
+        if (stepMetrics.getFailedEventsCount() > 0 && stopOnFailure) {
+            return JobStatus.STOP_ON_ERROR_PREFIX + step;
+        } else if( isStopped.get()){
+            return JobStatus.CANCELED_PREFIX + step;
+        } else if (stepMetrics.getFailedEventsCount() > 0 && stepMetrics.getSuccessfulEventsCount() > 0) {
+            return JobStatus.COMPLETED_WITH_ERRORS_PREFIX + step;
+        } else if (stepMetrics.getFailedEventsCount() == 0)  {
+            // Based on DHFPROD-5997, it is possible for a step to complete successfully but not process anything.
```

Comment:
I'm mostly wondering how safe this change is. When I first made it, the only test that failed was the one for an invalid sourceQuery. So I made the fix to CollectorImpl so that it throws an exception - its previous behavior for handling an error seemed incorrect.